actix-limitation/src/builder.rs

```diff
@@ -1,40 +1,60 @@
-use std::{borrow::Cow, time::Duration};
+use std::{borrow::Cow, sync::Arc, time::Duration};
 
+#[cfg(feature = "session")]
+use actix_session::SessionExt;
+use actix_web::dev::ServiceRequest;
 use redis::Client;
 
-use crate::{errors::Error, Limiter};
+use crate::{errors::Error, GetKeyFn, Limiter};
 
 /// Rate limiter builder.
 #[derive(Debug)]
 pub struct Builder {
     pub(crate) redis_url: String,
     pub(crate) limit: usize,
     pub(crate) period: Duration,
+    pub(crate) get_key_fn: Option<GetKeyFn>,
     pub(crate) cookie_name: Cow<'static, str>,
+    #[cfg(feature = "session")]
     pub(crate) session_key: Cow<'static, str>,
 }
 
 impl Builder {
     /// Set upper limit.
-    pub fn limit(&mut self, limit: usize) -> &mut Self {
+    pub fn limit(mut self, limit: usize) -> Self {
         self.limit = limit;
         self
     }
 
     /// Set limit window/period.
-    pub fn period(&mut self, period: Duration) -> &mut Self {
+    pub fn period(mut self, period: Duration) -> Self {
```

Comment:
This signature change is also breaking.

To be clear, i'm fine with breaks but just pointing it out given what you said. It's also a bit more common within the Actix ecosystem to use the `-> &mut Self` builder pattern (see `HttpResponseBuilder`).

Reply:
I reverted back to `&mut Self` 👍 
(now that it's wraped in Arc the closure is clonable and I can use an &mut self)

If you have the time would you mind explaining me why is Actix using `&mut Self` on builder pattern instead of `Self` ?

Reply:
It allows more options for configuring, both chaining and stand-alone mutation.

```
let limiter = Limiter::builder("redis://127.0.0.1").limt(500).build().unwrap();

// and

let mut limiter = Limiter::builder("redis://127.0.0.1");
limiter.limt(500);
let limiter = limiter.build().unwrap();
```

The latter being more difficult to integrate with conditional settings when using consuming builders.